nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -35,338 +35,406 @@
-: Node(name, name),
-  layered_costmap_(NULL),
-  name_(name),
-  tf_(tf),
-  transform_tolerance_(0.3),
-  map_update_thread_shutdown_(false),
-  stop_updates_(false),
-  initialized_(true),
-  stopped_(false),
-  map_update_thread_(NULL),
-  plugin_loader_("nav2_costmap_2d", "nav2_costmap_2d::Layer"),
-  publisher_(NULL),
-  last_publish_(0, 0, RCL_ROS_TIME),
-  publish_cycle_(1, 0),
-  footprint_padding_(0.0)
+Costmap2DROS::Costmap2DROS(const std::string & name)
+: nav2_lifecycle::LifecycleNode(name, "", true), name_(name)
 {
-  node_ = std::shared_ptr<rclcpp::Node>(this, [](rclcpp::Node *) {});
-
-  // Set Parameters if not set
-  set_parameter_if_not_set("transform_tolerance", 0.3);
-  set_parameter_if_not_set("update_frequency", 5.0);
-  set_parameter_if_not_set("publish_frequency", 1.0);
-  set_parameter_if_not_set("width", 10);
-  set_parameter_if_not_set("height", 10);
-  set_parameter_if_not_set("resolution", 0.1);
-  set_parameter_if_not_set("origin_x", 0.0);
-  set_parameter_if_not_set("origin_y", 0.0);
-  set_parameter_if_not_set("footprint", "[]");
-  set_parameter_if_not_set("footprint_padding", 0.01);
-  set_parameter_if_not_set("robot_radius", 0.1);
-
-  std::vector<std::string> plugin_names;
-  std::vector<std::string> plugin_types;
-  get_parameter_or_set("plugin_names", plugin_names,
-    {"static_layer", "obstacle_layer", "inflation_layer"});
-  get_parameter_or_set("plugin_types", plugin_types,
-    {"nav2_costmap_2d::StaticLayer",
-      "nav2_costmap_2d::ObstacleLayer",
-      "nav2_costmap_2d::InflationLayer"});
-
-  get_parameter_or<std::string>("global_frame", global_frame_, std::string("map"));
-  get_parameter_or<std::string>("robot_base_frame", robot_base_frame_, std::string("base_link"));
-
-  rclcpp::Time last_error = now();
-  std::string tf_error;
+  RCLCPP_INFO(get_logger(), "Creating");
+  client_node_ = std::make_shared<rclcpp::Node>(name + "_client");
+}
 
-  // we need to make sure that the transform between the robot base frame
-  // and the global frame is available
-  while (rclcpp::ok() &&
-    !tf_.canTransform(global_frame_, robot_base_frame_, tf2::TimePointZero,
-    tf2::durationFromSec(0.1), &tf_error))
-  {
-    if (last_error + nav2_util::duration_from_seconds(5.0) < now()) {
-      RCLCPP_WARN(
-        get_logger(),
-        "Timed out waiting for transform from %s to %s to become available before running costmap, tf error: %s", //NOLINT
-        robot_base_frame_.c_str(), global_frame_.c_str(), tf_error.c_str());
-      last_error = now();
-    }
-    // The error string will accumulate and errors will typically be the same, so the last
-    // will do for the warning above. Reset the string here to avoid accumulation.
-    tf_error.clear();
+Costmap2DROS::~Costmap2DROS()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+}
+
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+  getParameters();
```

Comment:
Should we be declaring and getting parameters in the configure step? It seems like we would be declaring parameters which have already been declared if someone unconfigures and then configures again the node, causing an error. I believe we discussed before that we should be declaring parameters in the constructor and then getting them again, `get_parameter`, in the configure step to avoid this issue. If I've got the idea wrong, let me know.

Reply:
No, we should be declaring them there. They should be declared in the constructor. Then, a node gets created once and its parameters declared once. I've moved the declarations to the constructor.